File: bms/go/geo/geo_s1_Interval_Expanded/geo_s1_Interval_Expanded.go
package test

import (
	"math"
)

var (
	geo_s1_dblEpsilon = 2.220446049e-16
)

func (i geo_s1_Interval) Expanded(margin float64) geo_s1_Interval {
	if margin >= 0 {
		if i.IsEmpty() {
			return i
		}
		// Check whether this interval will be full after expansion, allowing
		// for a rounding error when computing each endpoint.
		if i.Length()+2*margin+2*geo_s1_dblEpsilon >= 2*math.Pi {
			return geo_s1_FullInterval()
		}
	} else {
		if i.IsFull() {
			return i
		}
		// Check whether this interval will be empty after expansion, allowing
		// for a rounding error when computing each endpoint.
		if i.Length()+2*margin-2*geo_s1_dblEpsilon <= 0 {
			return geo_s1_EmptyInterval()
		}
	}
	result := geo_s1_IntervalFromEndpoints(
		math.Remainder(i.Lo-margin, 2*math.Pi),
		math.Remainder(i.Hi+margin, 2*math.Pi),
	)
	if result.Lo <= -math.Pi {
		result.Lo = math.Pi
	}
	return result
}

func (i geo_s1_Interval) IsEmpty() bool { return i.Lo == math.Pi && i.Hi == -math.Pi }

func geo_s1_IntervalFromEndpoints(lo, hi float64) geo_s1_Interval {
	i := geo_s1_Interval{lo, hi}
	if lo == -math.Pi && hi != math.Pi {
		i.Lo = math.Pi
	}
	if hi == -math.Pi && lo != math.Pi {
		i.Hi = math.Pi
	}
	return i
}

func (i geo_s1_Interval) IsFull() bool { return i.Lo == -math.Pi && i.Hi == math.Pi }

func (i geo_s1_Interval) Length() float64 {
	l := i.Hi - i.Lo
	if l >= 0 {
		return l
	}
	l += 2 * math.Pi
	if l > 0 {
		return l
	}
	return -1
}

func geo_s1_FullInterval() geo_s1_Interval { return geo_s1_Interval{-math.Pi, math.Pi} }

func geo_s1_EmptyInterval() geo_s1_Interval { return geo_s1_Interval{math.Pi, -math.Pi} }

type geo_s1_Interval struct {
	Lo, Hi float64
}

File: translations_only/fail/go/geo/geo_s1_Interval_Expanded/gemini/Restart/geo_s1_Interval_Expanded.rs

use std::f64::consts::PI;

const GEO_S1_DBL_EPSILON: f64 = 2.220446049e-16;

fn expanded(i: geo_s1_Interval, margin: f64) -> geo_s1_Interval {
    if margin >= 0.0 {
        if is_empty(&i) {
            return i;
        }
        // Check whether this interval will be full after expansion, allowing
        // for a rounding error when computing each endpoint.
        if length(&i) + 2.0 * margin + 2.0 * GEO_S1_DBL_EPSILON >= 2.0 * PI {
            return geo_s1_full_interval();
        }
    } else {
        if is_full(&i) {
            return i;
        }
        // Check whether this interval will be empty after expansion, allowing
        // for a rounding error when computing each endpoint.
        if length(&i) + 2.0 * margin - 2.0 * GEO_S1_DBL_EPSILON <= 0.0 {
            return geo_s1_empty_interval();
        }
    }
    let mut result = geo_s1_interval_from_endpoints(
        i.lo - margin,
        i.hi + margin,
    );
    if result.lo <= -PI {
        result.lo = PI;
    }
    return result;
}

fn is_empty(i: &geo_s1_Interval) -> bool {
    i.lo == PI && i.hi == -PI
}

fn geo_s1_interval_from_endpoints(lo: f64, hi: f64) -> geo_s1_Interval {
    let mut i = geo_s1_Interval { lo, hi };
    if lo == -PI && hi != PI {
        i.lo = PI;
    }
    if hi == -PI && lo != PI {
        i.hi = PI;
    }
    return i;
}

fn is_full(i: &geo_s1_Interval) -> bool {
    i.lo == -PI && i.hi == PI
}

fn length(i: &geo_s1_Interval) -> f64 {
    let l = i.hi - i.lo;
    if l >= 0.0 {
        return l;
    }
    let l = l + 2.0 * PI;
    if l > 0.0 {
        return l;
    }
    return -1.0;
}

fn geo_s1_full_interval() -> geo_s1_Interval {
    geo_s1_Interval {
        lo: -PI,
        hi: PI,
    }
}

fn geo_s1_empty_interval() -> geo_s1_Interval {
    geo_s1_Interval {
        lo: PI,
        hi: -PI,
    }
}

#[derive(Debug)]
struct geo_s1_Interval {
    lo: f64,
    hi: f64,
}
